### PR TITLE
fix(*) add flocks to additional services

### PR DESCRIPTION
**Pull request by jgmize**

### CODE REVIEW

**Comment by deis-admin** [comment]
Thanks for the contribution! Please ensure your commits follow our [style guide](http://docs.deis.io/en/latest/contributing/standards/#commit-style-guide). This code will be tested once a Deis maintainer reviews it.

Your Pull Request can earn you credit at DigitalOcean! See http://docs.deis.io/en/latest/contributing/hacking/#digitalocean-credit for details.

**Comment by krancour** [comment]
router and database could also need this.

**Comment by krancour** [comment]
Jenkins, add to whitelist.

**Comment by krancour** [comment]
These lines show components that are starting in parallel and may share layers beyond just alpine:

https://github.com/deis/deis/blob/master/deisctl/cmd/cmd.go#L156-L165

**Comment by carmstrong** [comment]
For reference, the root Docker issue is https://github.com/docker/docker/issues/9718. They claim it's fixed in master and will ship in Docker 1.9, but it would be great to confirm that.

**Comment by jgmize** [comment]
> router and database could also need this.

Updated to include flocks for router and database.

**Comment by krancour** [comment]
@jgmize, thank you for this!

@carmstrong one other thing we should consider... is _not_parallelizing the startup of these components in deisctl possibly _better_ than using more flock locks?  Or maybe should we do both?

**Comment by carmstrong** [comment]
> is not parallelizing the startup of these components in deisctl possibly better than using more flock locks

I think doing them in the units is probably safest, as anyone running our components would get that benefit. It's also easier to revert when (and if) we get this fixed in Docker. See also #4610

**Comment by carmstrong** [comment]
For builder and database, I'm wondering if we also need to lock the `*-data` containers as we did in #4590.

**Comment by jgmize** [comment]
> For builder and database, I'm wondering if we also need to lock the *-data containers as we did in #4590.

good catch, @carmstrong. I just added a commit to do that.

**Comment by krancour** [comment]
@jgmize we really appreciate this, but we're now getting reports of other problems with Docker 1.7.1 besides this.  We have a few balls in the air right now in terms of possibly rolling back to Docker 1.5 or shimming in a custom build of Docker.  One way or another, we won't end up using flock locks for this purpose-- and even the ones I had already added to logspout and publisher will certainly be removed as well.

Thanks again for the effort on this.

**Comment by mboersma** [comment]
#4614 is having some issues with `deis run`, so this `flock` approach may still be worth pursuing. I'm going to reopen it to see how it fares in CI--sorry for the whiplash!

File: deisctl/units/deis-builder.service [hunk 1]
```diff
@@ -4,8 +4,9 @@ Description=deis-builder
 [Service]
 EnvironmentFile=/etc/environment
 TimeoutStartSec=30m
-ExecStartPre=/bin/sh -c "docker inspect deis-builder-data >/dev/null 2>&1 || docker run --name deis-builder-data -v /var/lib/docker alpine:3.2 /bin/true"
-ExecStartPre=/bin/sh -c "IMAGE=`/run/deis/bin/get_image /deis/builder` && docker history $IMAGE >/dev/null 2>&1 || docker pull $IMAGE"
+ExecStartPre=/bin/sh -c "IMAGE=alpine:3.2 && docker history $IMAGE >/dev/null 2>&1 || flock -w 1200 /var/run/lock/alpine-pull docker pull $IMAGE"
+ExecStartPre=/bin/sh -c "IMAGE=alpine:3.2 && docker inspect deis-builder-data >/dev/null 2>&1 || docker run --name deis-builder-data -v /var/lib/docker $IMAGE /bin/true"
+ExecStartPre=/bin/sh -c "IMAGE=`/run/deis/bin/get_image /deis/builder` && docker history $IMAGE >/dev/null 2>&1 || flock -w 1200 /var/run/lock/alpine-pull docker pull $IMAGE"
```
Comment (bacongobbler) [inline deisctl/units/deis-builder.service]: maybe we should change `alpine-pull` to `builder-pull` here?

Comment (bacongobbler) [inline deisctl/units/deis-builder.service]: oh wait, I see the idea/usage behind `flock` now. Disregard.

**Comment by bacongobbler** [comment]
@krancour you did make a good point:

> These lines show components that are starting in parallel and may share layers beyond just alpine

I know we crafted a few lines in each component so that Jenkins can share these layers between images. I do agree that we share layers beyond alpine. We'll probably have to ensure that the images do **not** share layers somehow, or just `flock` every image pull, not just alpine. That means boot time will be significantly slower, though.

**Comment by mboersma** [comment]
> or just flock every image pull, not just alpine

This PR in conjunction with #4590 should mean that all `docker pull` operations in systemd units are guarded by `flock` now, with one lock for alpine-based images and one for ubuntu-based ones.

**Comment by bacongobbler** [comment]
Ah! Okay, maybe I was confused as to how `flock` worked. With `alpine-pull` being the lock name, I was confused that only the alpine image layers would be locked. I understand now :)

**Comment by bacongobbler** [comment]
manually tested, works as intended.

**Comment by krancour** [comment]
Thanks @jgmize!

**Pull request merged**